### PR TITLE
Pt 159406473 sc continuous chain monitoring 3

### PR DESCRIPTION
**Pull request by uwiger**
Rebased on top of latest changes.

Addresses:
* [#159406473 SC Watcher: listen for on-chain txs](https://www.pivotaltracker.com/story/show/159406473)
* [#159406302 SC WS: Solo closing](https://www.pivotaltracker.com/story/show/159406302)

Of particular note, the FSM now stays up during the closing phase, allowing the client to request `close_solo`, `slash` or `settle`. The FSM may also automatically slash if configured to do so (no tests exist for this). Documentation remains to be written. Also, http integration tests for the solo-closing phase.

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -89,7 +89,12 @@
 
         % waiting for https://github.com/szktty/erlang-lz4/pull/12 to be merged
         {lz4, {git, "https://github.com/aeternity/erlang-lz4.git",
-              {ref, "1ff9f36"}}}
+              {ref, "1ff9f36"}}},
+
+        % for lightweight trace-based debugging
+        {trace_runner, {git, "git://github.com/uwiger/trace_runner.git",
```
Comment (lucafavatella) [inline rebar.config]: Why not only in specific test profiles?

File: apps/aehttp/src/sc_ws_api_jsonrpc.erl [hunk 2]
```diff
@@ -493,6 +499,18 @@ process_request(#{<<"method">> := <<"channels.shutdown">>}, FsmPid) ->
     lager:warning("Channel WS: closing channel message received"),
     aesc_fsm:shutdown(FsmPid),
     no_reply;
+process_request(#{<<"method">> := <<"channels.close_solo">>}, FsmPid) ->
```
Comment (lucafavatella) [inline apps/aehttp/src/sc_ws_api_jsonrpc.erl]: Does this require update to `node/api/channel_ws_api.md` in protocol repo (after https://github.com/aeternity/protocol/pull/328 merged )?

Comment (uwiger) [reply apps/aehttp/src/sc_ws_api_jsonrpc.erl]: Yes, will update the documentation

**Comment by lucafavatella** [comment]
If superseded by this, could you please consider closing https://github.com/aeternity/aeternity/pull/2344 and https://github.com/aeternity/aeternity/pull/2378 ?

File: apps/aechannel/src/aesc_fsm.erl [hunk 3]
```diff
@@ -1962,6 +2152,30 @@ slash_tx(Account, Nonce, _Round, SignedTx, #data{ on_chain_id = ChanId
...
     {ok, Tx, []}.
 
+settle_tx_for_signing(#data{} = D) ->
+    Account = my_account(D),
+    {ok, Nonce} = aec_next_nonce:pick_for_account(Account),
+    settle_tx(Account, Nonce, D).
+
+settle_tx(Account, Nonce, #data{ opts  = #{initiator := I,
+                                           responder := R} = Opts
+                               , state = State } = D) ->
```
Comment (velzevur) [inline apps/aechannel/src/aesc_fsm.erl]: I see the settle tx depends heavily on the latest state. Although this could be fragile with the play of Force Progress on-chain.
This is outside of the scope of this PR, in order for us not to lose it - tracked as [SC: Make settling force-progress aware](https://www.pivotaltracker.com/story/show/165977998)

Comment (velzevur) [reply apps/aechannel/src/aesc_fsm.erl]: As same reasoning applies to solo close as well, I extended the PT

**Comment by uwiger** [comment]
> I see this PR does multiple useful things:
> 
>     * Adds development script (`scripts/log_xtract.escript`) for easing inspecting logs.

[#165979853](https://www.pivotaltracker.com/story/show/165979853)

>     * Adds framework for tracing tests (I assume it is independent from the rest).

[#165980068](https://www.pivotaltracker.com/story/show/165980068)

>     * Refactors types and exports in channel related modules.

[#165980068](https://www.pivotaltracker.com/story/show/165980068)
> 
>     * Adds debug logging re block insertion time.

This we can probably do without.

>     * Enables usage of channel solo close from websocket API.
> 
>     * Enhances internal of channel client FSM in order to do the above (i.e. enabling usage of channel solo close from websocket API).

These two I don't know if I can split easily.

File: rebar.config (hunk 1 above)
Comment (uwiger) [reply rebar.config]: @lucafavatella Because the code needs to be instrumented (ideally using the `event/1` function rather than debug logging, which is supported by the `trace_runner.hrl`.)

While it is possible to instrument the code without having `trace_runner` included, having it included also makes it possible for dialyzer to check the instrumentation callbacks.

**Comment by uwiger** [comment]
Superseded by #2431 - closing this PR
